### ISSUE TITLE
NFC driver not freeing /dev/serial0 on authentication failure?

### ISSUE THREAD

**Issue by fussel132** [kormax/apple-home-key-reader#3]
Hi,
first of all: FINALLY. Idk when homekeys were announced but it is very great to see an implementation like this.

I've deployed the project on a Raspberry Pi and it is running fine. However, in some cases the loop service.py/_process_nfc() throws a TimeoutError:

<details>
  <summary>View Log</summary>
  
  >[2023-12-05 10:48:54,116] [   ERROR] service           :156  [Errno 110] Connection timed out
  Traceback (most recent call last):
    File "/home/pi/apple-home-key-reader/service.py", line 154, in run
      self._process_nfc()
    File "/home/pi/apple-home-key-reader/service.py", line 95, in _process_nfc
      remote_target = self.clf.sense(
    File "/home/pi/apple-home-key-reader/util/bfclf.py", line 162, in sense
      sense_broadcast(target, options.get("broadcast", None))
    File "/home/pi/apple-home-key-reader/util/bfclf.py", line 124, in sense_broadcast
      _ = self.device.chipset.in_communicate_thru(broadcast, timeout=0.1)
    File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn53x.py", line 416, in in_communicate_thru
      data = self.command(0x42, data, timeout)
    File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn53x.py", line 202, in command
      raise error
    File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn53x.py", line 197, in command
      frame = self.read_frame(int(1000 * timeout))
    File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn53x.py", line 249, in read_frame
      return self.transport.read(timeout)
    File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/transport.py", line 154, in read
      raise IOError(errno.ETIMEDOUT, os.strerror(errno.ETIMEDOUT))
  TimeoutError: [Errno 110] Connection timed out
</details>

I've observed this behaviour in three scenarios now:

1. Running application, then presenting a static NFC tag resulting in an unrecoverable timeout error. After the delay, the terminal gets spammed with the TimeoutError as seen above.

<details>
  <summary>View Log</summary>
  
  >[2023-12-05 10:57:54,211] [    INFO] service           :147  Waiting for next device...
  [2023-12-05 10:57:54,242] [    INFO] service           :110  Got NFC tag Type2Tag ID=A3031BF7
  [2023-12-05 10:57:54,243] [    INFO] homekey           :529  SELECT CMD = ISO7816Command(cla=0x00; ins=0xa4; p1=0x04; p2=0x00; lc=0x07(7); data=a0000008580101; le=0x00)
  [2023-12-05 10:57:54,581] [   ERROR] service           :142  unrecoverable timeout error
  Traceback (most recent call last):
    File "/home/pi/apple-home-key-reader/service.py", line 112, in _process_nfc
      result_flow, new_issuers_state, endpoint = read_homekey(
    File "/home/pi/apple-home-key-reader/homekey.py", line 706, in read_homekey
      response = select_applet(tag, applet=ISO7816Application.HOME_KEY)
    File "/home/pi/apple-home-key-reader/homekey.py", line 530, in select_applet
      response = tag.transceive(command)
    File "/home/pi/apple-home-key-reader/util/iso7816.py", line 199, in transceive
      self._implementation.transceive(
    File "/home/pi/.local/lib/python3.9/site-packages/nfc/tag/tt2.py", line 600, in transceive
      raise Type2TagCommandError(nfc.tag.TIMEOUT_ERROR)
  nfc.tag.tt2.Type2TagCommandError: unrecoverable timeout error
  [2023-12-05 10:57:54,586] [ WARNING] service           :143  Encountered an exception. Waiting for 5 seconds before continuing...
  [2023-12-05 10:57:59,592] [    INFO] service           :147  Waiting for next device...
  [2023-12-05 10:57:59,723] [   ERROR] service           :156  [Errno 110] Connection timed out
  Traceback (most recent call last):
    File "/home/pi/apple-home-key-reader/service.py", line 154, in run
      self._process_nfc()
    File "/home/pi/apple-home-key-reader/service.py", line 95, in _process_nfc
      remote_target = self.clf.sense(
    File "/home/pi/apple-home-key-reader/util/bfclf.py", line 162, in sense
      sense_broadcast(target, options.get("broadcast", None))
    File "/home/pi/apple-home-key-reader/util/bfclf.py", line 124, in sense_broadcast
      _ = self.device.chipset.in_communicate_thru(broadcast, timeout=0.1)
    File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn53x.py", line 416, in in_communicate_thru
      data = self.command(0x42, data, timeout)
    File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn53x.py", line 202, in command
      raise error
    File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn53x.py", line 197, in command
      frame = self.read_frame(int(1000 * timeout))
    File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn53x.py", line 249, in read_frame
      return self.transport.read(timeout)
    File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/transport.py", line 154, in read
      raise IOError(errno.ETIMEDOUT, os.strerror(errno.ETIMEDOUT))
  TimeoutError: [Errno 110] Connection timed out
  [...and so on...]
</details>

2. Running application (no matter how long or how many authentication attempts of any type), then killing it (ctrl + c). Trying to restart the application (even after several minutes) won't work due to (again) the same TimeoutError killing the app right at the beginning. The last log message looks like this:
> [2023-12-05 10:58:03,746] [   ERROR] base_events       :1738 Task was destroyed but it is pending!
task: <Task pending name='Task-11' coro=<Zeroconf._async_broadcast_service() done, defined at /home/pi/.local/lib/python3.9/site-packages/zeroconf/_core.py:367> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f8cb87dc0>()]>>

But I don't think thats the task keeping the reader busy as zeroconf is some mDNS stuff I guess.
<details>
  <summary>View Full Log</summary>

> [...]
^C[2023-12-05 10:58:03,558] [    INFO] main              :71   SIGNAL 15
[2023-12-05 10:58:03,680] [    INFO] accessory_driver  :432  Stopping accessory NFC Lock on address ['xxx.xxx.xxx.xxx'], port 51926.
[2023-12-05 10:58:03,683] [    INFO] accessory_driver  :342  Closed the event loop
[2023-12-05 10:58:03,746] [   ERROR] base_events       :1738 Task was destroyed but it is pending!
task: <Task pending name='Task-11' coro=<Zeroconf._async_broadcast_service() done, defined at /home/pi/.local/lib/python3.9/site-packages/zeroconf/_core.py:367> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f8cb87dc0>()]>>
pi@nfctest:\~/apple-home-key-reader $ python3 main.py
[2023-12-05 11:07:03,300] [    INFO] __init__          :148  searching for reader on path tty:serial0:pn532
Traceback (most recent call last):
  File "/home/pi/apple-home-key-reader/main.py", line 82, in <module>
    main()
  File "/home/pi/apple-home-key-reader/main.py", line 63, in main
    nfc_device = configure_nfc_device(config["nfc"])
  File "/home/pi/apple-home-key-reader/main.py", line 41, in configure_nfc_device
    clf = BroadcastFrameContactlessFrontend(
  File "/home/pi/apple-home-key-reader/util/bfclf.py", line 51, in __init__
    super().__init__(path)
  File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/__init__.py", line 75, in __init__
    if path and not self.open(path):
  File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/__init__.py", line 149, in open
    self.device = device.connect(path)
  File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/device.py", line 112, in connect
    device = driver.init(tty)
  File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn532.py", line 411, in init
    if not transport.read(timeout=initial_timeout) == Chipset.ACK:
  File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/transport.py", line 154, in read
    raise IOError(errno.ETIMEDOUT, os.strerror(errno.ETIMEDOUT))
TimeoutError: [Errno 110] Connection timed out
pi@nfctest:\~/apple-home-key-reader $ 
</details>

3. When unplugging the reader while running. It will result in another spam of the terminal. Sadly, it is not recovering from that when you reconnect the reader again :(

<details>
  <summary>View Log</summary>

>[2023-12-05 11:15:55,207] [   ERROR] pn53x             :180  input/output error while waiting for ack
[2023-12-05 11:15:55,208] [   ERROR] service           :156  [Errno 5] Input/output error
Traceback (most recent call last):
  File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn53x.py", line 178, in command
    frame = self.read_frame(timeout=100)
  File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn53x.py", line 249, in read_frame
    return self.transport.read(timeout)
  File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/transport.py", line 154, in read
    raise IOError(errno.ETIMEDOUT, os.strerror(errno.ETIMEDOUT))
TimeoutError: [Errno 110] Connection timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pi/apple-home-key-reader/service.py", line 154, in run
    self._process_nfc()
  File "/home/pi/apple-home-key-reader/service.py", line 95, in _process_nfc
    remote_target = self.clf.sense(
  File "/home/pi/apple-home-key-reader/util/bfclf.py", line 142, in sense
    self.device.mute()  # deactivate the rf field
  File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn53x.py", line 499, in mute
    self.chipset.rf_configuration(0x01, bytearray([0b00000010]))
  File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn53x.py", line 348, in rf_configuration
    self.command(0x32, bytearray([cfg_item]) + bytearray(cfg_data),
  File "/home/pi/.local/lib/python3.9/site-packages/nfc/clf/pn53x.py", line 181, in command
    raise IOError(errno.EIO, os.strerror(errno.EIO))
OSError: [Errno 5] Input/output error
[... repeating over and over...]
</details>

So in conclusion, it seems that the connection to the reader via UART is never correctly closed. There is no other pyhon process running in the background after killing the code but interpreting those error messages it appears to me that the application cannot access the reader. Because to fix ANY of the above named scenarios: **Unplug and replug the reader, then restart**

I've digged a little bit into the code but wasn't able to find a quick fix for the two problems (timeout and i/o errors)

Maybe python 3.9.2 is not compatible?

**Comment by kormax** [comment]
Hello. 
Thanks for sharing this info.

Here are a couple of mine comments (not in same order, numbers just for formatting reasons):

1. I indeed didn't test shutdown handling for the application too much, so it's possible that it exits with exceptions.

2. From my experience, not closing the port is not an issue if application has been fully closed (process killed), as that's something the OS takes care of (no process to hold to FD. 

3. The issue with app breaking after type2 tag use was reported by one of my testers, but I wasn't able to recreate it. I'll try to take a look, as it seems serious enough. I'll report back.

4. The app indeed doesn't re-open the device if connection closes. I'll think about moving nfc clf creation logic inside the loop. That would be useful.

**Comment by kormax** [comment]
💡Issues with not being open the device after a force application restart are related to the fact that nfcpy changes baud rate of the device if the OS is linux. In case of an abrupt shutdown it doesn't change it back, and upon next initialisation it only tries the standard 115200, so it fails. Reconnecting the module fixes the issue because it resets back the baud rate configuration.

**Comment by fussel132** [comment]
So in the sigterm event we need to tell the PN532 to switch back to it's original baud rate or is there even a reset command we can send to the reader?

**Comment by kormax** [comment]
I think nfcpy already has the needed code in `close()` method of the clf. It can be implemented via signal handling, but I personally wouldn't pursue adding it.

Thing is, even if close() call is added to shutdown handling, that code is not guaranteed to run in case of a random system crash or restart, so there's still no guarantee that the pn532 resets the baud rate. 
Most guaranteed way of ensuring stability is either 1) removing baud rate switching; 2) adding code for searching the needed baud rate.

If you really want to do something about this issue right now, I would modify a local version of nfpy [on this exact line and file](https://github.com/nfcpy/nfcpy/blob/8a195262fe86167845d282ba5bc9cdaadf6441c4/src/nfc/clf/pn532.py#L391), by adding `False and` to make that baudrate-switching code not run.

**Comment by fussel132** [comment]
Okay and about the baud rate changes you are absolutely right. Using nfcpy close() method resets the baud rate to 115200. So to make this failproof, we'd need to cap the max_baudrate at 115200 and as far as I see that is only possible by hardcoding it into nfcpy.
Your link to where the `False` needs to be is off by one line. [This one](https://github.com/nfcpy/nfcpy/blob/master/src/nfc/clf/pn532.py#L426) (one before) is correct.
A solution to this at nfcpy is pending already, but it hasn't made progress in the last two years: https://github.com/nfcpy/nfcpy/pull/191

**Comment by kormax** [comment]
@fussel132 needed changes integrated into main. Please, try them out≥

**Comment by fussel132** [comment]
Just did so, works GREAT. Way better than my suggestion. Out of curiosity, the crash that appeared when presenting a type2 tag, what specifically was the problem there? My solution by just reconnecting the reader in software happened right away. The latest additions still tell me to remove the target from the field :)

Hotplugging works great! Restart problem was fixed by keeping the baud rate at 115200, but as you said previously that's done in nfcpy.

**Comment by kormax** [comment]
Idk actually what caused the issue with t2 tags in the first place.

I assume the problem was a combo of issues caused by pn532 (which handles lower levels of nfc stack by itself) my broadcast frames mod (which sent commands while bypassing pn532 internal stack), and perhaps the original nfcpy driver impl.

I've just verified that it works this way so decided to leave it as is.

**Comment by kormax** [comment]
Hello.

[This PR](https://github.com/kormax/apple-home-key-reader/pull/7) should be able to fix baudrate negotiation issue for everyone automatically (by monkey-patching nfcpy library code at runtime to disable baudrate change). 

If you have some time, I would be glad if you could try it out. With this change manual changes to nfcpy would no longer be necessary.
